react/Gallery.js

```diff
@@ -59,62 +62,81 @@ const Gallery = ({
       >
-        {({ measure }) => (
-          <div className={containerClasses} key={key} style={style} onLoad={measure}>
-            {rowItems.map(renderItem)}
-          </div>
-        )}
+        <div className={containerClasses} key={key} style={style}>
+          {rowItems.map(this.renderItem)}
+        </div>
       </CellMeasurer>
     )
   }
 
-  const ssrContainer = classNames(`${searchResult.gallery} pa3 bn`, {
-    [searchResult.galleryTwoColumns]: layoutMode == 'small' && mobile
-  })
+  handleSSR() {
+    const { products, layoutMode, runtime: { hints: { mobile } } } = this.props
 
-  const ssrFallBack = (
-    <div className={ssrContainer}>
-      {products.map(renderItem)}
-    </div>
-  )
-
-  return (
-    <NoSSR onSSR={ssrFallBack}>
-      <div className={`${searchResult.gallery} pa3 bn`}>
-        <Adopt
-          mapper={{
-            scroller: <WindowScroller />,
-            autoSizer: <AutoSizer disableHeight />,
-          }}
-          mapProps={({ scroller: { height }, autoSizer: { width } }) => ({ width, height })}
-        >
-          {({ width, height }) => {
-            const itemsPerRow = (layoutMode === 'small' && mobile) ? TWO_ITEMS : (Math.floor(width / itemWidth) || ONE_ITEM)
-            const nRows = Math.ceil(products.length / itemsPerRow)
-
-            return (
-              <List
-                key={layoutMode}
-                deferredMeasurementCache={cache}
-                autoHeight
-                width={width}
-                height={height}
-                rowCount={nRows}
-                rowHeight={cache.rowHeight}
-                rowRenderer={args => renderRow({ ...args, itemsPerRow })}
-              />
-            )
-          }}
-        </Adopt>
-      </div >
-    </NoSSR>
-  )
+    const ssrContainer = classNames(`${searchResult.gallery} pa3 bn`, {
+      [searchResult.galleryTwoColumns]: layoutMode === 'small' && mobile,
+    })
+
+    return (
+      <div className={ssrContainer}>
+        {products.map(this.renderItem)}
+      </div>
+    )
+  }
+
+  render() {
+    const {
+      products,
+      layoutMode,
+      runtime: { hints: { mobile } },
+      itemWidth,
+    } = this.props
+
+    // Reset the cache to recalculate the heights when layout changes
+    if (layoutMode !== this.state.prevLayoutMode) {
+      this.cache.clearAll()
+    }
+
+    return (
+      <NoSSR onSSR={this.handleSSR()}>
+        <div className={`${searchResult.gallery} pa3 bn`}>
+          <Adopt
+            mapper={{
+              scroller: <WindowScroller />,
+              autoSizer: <AutoSizer disableHeight />,
+            }}
+            mapProps={({ scroller: { height, scrollTop, isScrolling }, autoSizer: { width } }) => ({ width, height, scrollTop, isScrolling })}
+          >
+            {({ width, height, scrollTop, isScrolling }) => {
+              const itemsPerRow = (layoutMode === 'small' && mobile) ? TWO_ITEMS : (Math.floor(width / itemWidth) || ONE_ITEM)
```

Comment:
You could move this to outside `return` block for better readability, what do you think?

Reply:
Render would be like that:
```js
render() {
    const {
      layoutMode,
    } = this.props

    // Reset the cache to recalculate the heights when layout changes
    if (layoutMode !== this.state.prevLayoutMode) {
      this.setState({ layoutMode })
      this.cache.clearAll()
    }

    const mapContainerProps = ({ scroller: { height, scrollTop, isScrolling }, autoSizer: { width } }) => ({ width, height, scrollTop, isScrolling })

    return (
      <NoSSR onSSR={this.handleSSR()}>
        <div className={`${searchResult.gallery} pa3 bn`}>
          <Adopt
            mapper={{
              scroller: <WindowScroller />,
              autoSizer: <AutoSizer disableHeight />,
            }}
            mapProps={mapContainerProps}
          >
            {this.renderList}
          </Adopt>
        </div >
      </NoSSR>
    )
  }
```
I think that doing it in that way separates related concepts and makes the render props harder to see. Also, I would be repeating some for for destructuring in renderList.

Reply:
I prefer this solution, but anyway i will pass it and can be discussed forwardly